docker/lr-pb/shard_bam.py

```diff
@@ -0,0 +1,124 @@
+
+    print(f"Reading index ({pbi}). This may take a few minutes...", flush=True)
+
+    # Decode PacBio .pbi file.  This is not a full decode of the index, only the parts we need for sharding.
+    # More on index format at https://pacbiofileformats.readthedocs.io/en/9.0/PacBioBamIndex.html .
+
+    fmt = Struct(
+        # Header
+        "magic" / Const(b"PBI\x01"),
+        "version_patch" / Int8ul,
+        "version_minor" / Int8ul,
+        "version_major" / Int8ul,
+        "version_empty" / Int8ul,
+        "pbi_flags" / Int16ul,
+        "n_reads" / Int32ul,
+        "reserved" / Padding(18),
+
+        # Basic information section (columnar format)
+        # "rgId" / LazyArray(this.n_reads, Int32sl),
+        # "qStart" / LazyArray(this.n_reads, Int32sl),
+        # "qEnd" / LazyArray(this.n_reads, Int32sl),
+        # "holeNumber" / Array(this.n_reads, Int32sl),
+        # "readQual" / LazyArray(this.n_reads, Float32b),
+        # "ctxtFlag" / LazyArray(this.n_reads, Int8ul),
+        # "fileOffset" / Array(this.n_reads, Int64sl),
+
+        "rgId" / Padding(this.n_reads * 4),
+        "qStart" / Padding(this.n_reads * 4),
+        "qEnd" / Padding(this.n_reads * 4),
+        "holeNumber" / Array(this.n_reads, Int32sl),
+        "readQual" / Padding(this.n_reads * 4),
+        "ctxtFlag" / Padding(this.n_reads * 1),
+        "fileOffset" / Array(this.n_reads, Int64sl),
+    )
+
+    # Make a list of bgzf virtual file offsets for sharding.
+    file_offsets_hash = OrderedDict()
+
+    last_offset = 0
+    with gzip.open(pbi, "rb") as f:
+        a = fmt.parse_stream(f)
+
+        for i in range(0, a.n_reads):
+            # Save only the virtual file offset for the first ZMW hole number, so that shard boundaries
+            # always keep reads from the same ZMW together.
+            if a.holeNumber[i] not in file_offsets_hash:
+                file_offsets_hash[a.holeNumber[i]] = a.fileOffset[i]
+
+            last_offset = a.fileOffset[i]
+
+    file_offsets = list(file_offsets_hash.values())
+
+    shard_offsets = []
+    for i in range(0, len(file_offsets), ceil(len(file_offsets)/args.num_shards)):
+        shard_offsets.append(file_offsets[i])
+
+    # For the last read in the file, pad the offset so the final comparison in write_shard() retains the final read.
+    shard_offsets.append(last_offset + 100)
+
+    # Prepare a function with arguments partially filled in (the later imap_unordered() call requires functions that
+    # only have one remaining argument to be specified).
+    func = partial(write_shard, args.bam, shard_offsets, args.prefix)
+    idx = list(range(0, len(shard_offsets) - 1))
+
+    # Write the shards using the specified number of threads.
+    print(f"Writing {len(idx)} shards using {args.num_threads} threads...", flush=True)
+    res = ThreadPool(args.num_threads).imap_unordered(func, idx)
+
+    # Emit final stats on the sharding.
+    all_num_reads_written = list(res)
```

Comment:
could you make it a bit more specific, i.e. checking that each ZMW has the same number of reads before and after sharding? That'll convince me that this does shard all subreads of the same ZMW into one sub-bam.

Reply:
seems like still a check on total number of reads?

Reply:
what I meant was to have a check such that each ZMW have the same number of reads in the original BAM and and the sharded BAM, e.g. having a dictionary each before and after, and make sure they are equal.

Reply:
Already done - it's just in two different places.  The per-ZMW check is in the write_shard function (lines 41-42).  The check on line 135 is to ensure that all threads have completed successfully and the total number of reads they wrote adds up to the total number of reads in the BAM file.

Reply:
Now that there's a test specific to this script, I think these checks should be removed.